File: sdk/go/provider/m2u.go
package provider

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/url"
	"strings"

	http "github.com/bogdanfinn/fhttp"
)

const m2uBase = "https://api.m2u.io"

type m2uMailboxResponse struct {
	Mailbox struct {
		Token     string `json:"token"`
		ViewToken string `json:"view_token"`
		LocalPart string `json:"local_part"`
		Domain    string `json:"domain"`
		ExpiresAt any    `json:"expires_at"`
		CreatedAt any    `json:"created_at"`
	} `json:"mailbox"`
}

type m2uListResponse struct {
	Messages []map[string]any `json:"messages"`
}

type m2uDetailResponse struct {
	Message map[string]any `json:"message"`
}

func m2uJSON(method, u string, body []byte, out any) error {
	var reader io.Reader
	if body != nil {
		reader = bytes.NewReader(body)
	}
	req, err := http.NewRequest(method, u, reader)
	if err != nil {
		return err
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36")
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}

	resp, err := HTTPClient().Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("m2u: http %d", resp.StatusCode)
	}
	if out == nil {
		return nil
	}
	return json.Unmarshal(raw, out)
}

func m2uAnyString(v any) string {
	if v == nil {
		return ""
	}
	switch x := v.(type) {
	case string:
		return strings.TrimSpace(x)
	default:
		return strings.TrimSpace(fmt.Sprint(v))
	}
}

func m2uPackToken(token, viewToken string) string {
	raw, _ := json.Marshal(map[string]string{
		"token":     token,
		"viewToken": viewToken,
	})
	return string(raw)
}

func m2uUnpackToken(packed string) (string, string) {
	value := strings.TrimSpace(packed)
	if value == "" {
		return "", ""
	}
	if strings.HasPrefix(value, "{") {
		var payload map[string]any
		if err := json.Unmarshal([]byte(value), &payload); err == nil {
			return m2uAnyString(payload["token"]), m2uAnyString(payload["viewToken"])
		}
		return "", ""
	}
	return value, ""
}

func m2uFlattenMessage(raw map[string]any, recipient string) map[string]any {
	id := raw["id"]
	if m2uAnyString(id) == "" {
		id = raw["message_id"]
	}
	from := raw["from_addr"]
	if m2uAnyString(from) == "" {
		from = raw["from_address"]
	}
	if m2uAnyString(from) == "" {
		from = raw["from"]
	}
	text := raw["text_body"]
	if m2uAnyString(text) == "" {
		text = raw["body_text"]
	}
	if m2uAnyString(text) == "" {
		text = raw["text"]
	}
	html := raw["html_body"]
	if m2uAnyString(html) == "" {
		html = raw["body_html"]
	}
	if m2uAnyString(html) == "" {
		html = raw["html"]
	}
	date := raw["received_at"]
	if m2uAnyString(date) == "" {
		date = raw["created_at"]
	}
	return map[string]any{
		"id":          id,
		"from":        from,
		"to":          recipient,
		"subject":     raw["subject"],
		"text":        text,
		"html":        html,
		"date":        date,
		"is_read":     raw["is_read"],
		"attachments": raw["attachments"],
	}
}

func m2uFetchDetail(token, viewToken, messageID string) map[string]any {
	u := fmt.Sprintf("%s/v1/mailboxes/%s/messages/%s?view=%s", m2uBase, url.PathEscape(token), url.PathEscape(messageID), url.QueryEscape(viewToken))
	var data m2uDetailResponse
	if err := m2uJSON(http.MethodGet, u, nil, &data); err != nil {
		return nil
	}
	if data.Message == nil {
		return nil
	}
	return data.Message
}

func M2uGenerate() (*CreatedMailbox, error) {
	var data m2uMailboxResponse
	if err := m2uJSON(http.MethodPost, m2uBase+"/v1/mailboxes/auto", []byte("{}"), &data); err != nil {
		return nil, err
	}
	localPart := strings.TrimSpace(data.Mailbox.LocalPart)
	domain := strings.TrimSpace(data.Mailbox.Domain)
	token := strings.TrimSpace(data.Mailbox.Token)
	viewToken := strings.TrimSpace(data.Mailbox.ViewToken)
	email := ""
	if localPart != "" && domain != "" {
		email = localPart + "@" + domain
	}
	if email == "" || token == "" || viewToken == "" {
		return nil, fmt.Errorf("m2u: invalid mailbox response")
	}
	return &CreatedMailbox{
		Channel:   "m2u",
		Email:     email,
		Token:     m2uPackToken(token, viewToken),
		ExpiresAt: data.Mailbox.ExpiresAt,
		CreatedAt: m2uAnyString(data.Mailbox.CreatedAt),
	}, nil
}

func M2uGetEmails(token, email string) ([]NormEmail, error) {
	mailboxToken, viewToken := m2uUnpackToken(token)
	address := strings.TrimSpace(email)
	if mailboxToken == "" {
		return nil, fmt.Errorf("m2u: missing token")
	}
	if viewToken == "" {
		return nil, fmt.Errorf("m2u: missing view token")
	}
	if address == "" {
		return nil, fmt.Errorf("m2u: empty email")
	}

	u := fmt.Sprintf("%s/v1/mailboxes/%s/messages?view=%s", m2uBase, url.PathEscape(mailboxToken), url.QueryEscape(viewToken))
	var data m2uListResponse
	if err := m2uJSON(http.MethodGet, u, nil, &data); err != nil {
		return nil, err
	}
	out := make([]NormEmail, 0, len(data.Messages))
	for _, row := range data.Messages {
		raw := row
		if id := m2uAnyString(row["id"]); id != "" {
			if detail := m2uFetchDetail(mailboxToken, viewToken, id); detail != nil {
				raw = detail
			}
		}
		out = append(out, NormalizeMap(m2uFlattenMessage(raw, address), address))
	}
	return out, nil
}

File: sdk/npm/src/providers/m2u.ts
import { InternalEmailInfo, Email, Channel } from '../types';
import { normalizeEmail } from '../normalize';
import { fetchWithTimeout } from '../retry';

const CHANNEL: Channel = 'm2u';
const API_BASE = 'https://api.m2u.io';

type Mailbox = {
  token?: string;
  view_token?: string;
  local_part?: string;
  domain?: string;
  expires_at?: string | number;
  created_at?: string | number;
};

type CreateResponse = {
  mailbox?: Mailbox;
};

type ListResponse = {
  messages?: Array<Record<string, unknown>>;
};

type DetailResponse = {
  message?: Record<string, unknown>;
};

function headers(extra?: Record<string, string>): Record<string, string> {
  return {
    Accept: 'application/json',
    'User-Agent':
      'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36',
    ...(extra || {}),
  };
}

async function fetchJson<T>(url: string, init?: RequestInit): Promise<T> {
  const response = await fetchWithTimeout(url, {
    ...init,
    headers: {
      ...headers(),
      ...(init?.headers || {}),
    },
  });
  if (!response.ok) {
    throw new Error(`m2u http ${response.status}`);
  }
  return response.json() as Promise<T>;
}

function packToken(token: string, viewToken: string): string {
  return JSON.stringify({ token, viewToken });
}

function unpackToken(packed: string): { token: string; viewToken: string } {
  const value = String(packed || '').trim();
  if (!value) {
    return { token: '', viewToken: '' };
  }
  if (value.startsWith('{')) {
    try {
      const data = JSON.parse(value) as Record<string, unknown>;
      return {
        token: typeof data.token === 'string' ? data.token.trim() : '',
        viewToken: typeof data.viewToken === 'string' ? data.viewToken.trim() : '',
      };
    } catch {
      return { token: '', viewToken: '' };
    }
  }
  return { token: value, viewToken: '' };
}

function flattenMessage(raw: Record<string, unknown>, recipient: string): Record<string, unknown> {
  return {
    id: raw.id || raw.message_id,
    from: raw.from_addr || raw.from_address || raw.from || '',
    to: recipient,
    subject: raw.subject || '',
    text: raw.text_body || raw.body_text || raw.text || '',
    html: raw.html_body || raw.body_html || raw.html || '',
    date: raw.received_at || raw.created_at || '',
    isRead: raw.is_read ?? raw.isRead ?? raw.seen ?? false,
    attachments: raw.attachments || [],
  };
}

export async function generateEmail(): Promise<InternalEmailInfo> {
  const data = await fetchJson<CreateResponse>(`${API_BASE}/v1/mailboxes/auto`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({}),
  });
  const mailbox = data.mailbox || {};
  const localPart = String(mailbox.local_part || '').trim();
  const domain = String(mailbox.domain || '').trim();
  const token = String(mailbox.token || '').trim();
  const viewToken = String(mailbox.view_token || '').trim();
  const email = localPart && domain ? `${localPart}@${domain}` : '';
  if (!email || !token || !viewToken) {
    throw new Error('m2u: invalid mailbox response');
  }
  return {
    channel: CHANNEL,
    email,
    token: packToken(token, viewToken),
    expiresAt: mailbox.expires_at,
    createdAt: mailbox.created_at == null ? undefined : String(mailbox.created_at),
  };
}

async function fetchDetail(token: string, viewToken: string, messageId: string): Promise<Record<string, unknown> | null> {
  try {
    const data = await fetchJson<DetailResponse>(
      `${API_BASE}/v1/mailboxes/${encodeURIComponent(token)}/messages/${encodeURIComponent(messageId)}?view=${encodeURIComponent(viewToken)}`,
    );
    return data.message && typeof data.message === 'object' ? data.message : null;
  } catch {
    return null;
  }
}

export async function getEmails(token: string, email: string): Promise<Email[]> {
  const { token: mailboxToken, viewToken } = unpackToken(token);
  const address = String(email || '').trim();
  if (!mailboxToken) throw new Error('m2u: missing token');
  if (!viewToken) throw new Error('m2u: missing view token');
  if (!address) throw new Error('m2u: empty email');

  const data = await fetchJson<ListResponse>(
    `${API_BASE}/v1/mailboxes/${encodeURIComponent(mailboxToken)}/messages?view=${encodeURIComponent(viewToken)}`,
  );
  const rows = Array.isArray(data.messages) ? data.messages : [];
  const emails: Email[] = [];
  for (const row of rows) {
    const id = String(row.id || row.message_id || '').trim();
    const detail = id ? await fetchDetail(mailboxToken, viewToken, id) : null;
    emails.push(normalizeEmail(flattenMessage(detail || row, address), address));
  }
  return emails;
}
